### ISSUE TITLE
mxnet multicore on LInux in R

### ISSUE THREAD

**Issue by shivonkar** [apache/incubator-mxnet#8575]
Follwoing works fine on windows although throws error on Linux

      core <- 5
      ctx <- lapply(c(1:core), function(i) {mx.cpu(i)})

The model function is "mx.model.FeedForward.create"

       The error is "Error in (function (symbol, ctx, grad.req = "null", ...) : Need more shape information to decide the shapes of arguments"

Although, it works properly when core <- 1

**Comment by thirdwing** [comment]
Please provide a reproducible example.

**Comment by shivonkar** [comment]
The working test code is here and error is attached

g_working_directory <- ifelse(Sys.info()['sysname'] == "Windows",  "D:/INFY_BACKUP/Projects/Analytics/Practice/R/Prospects/amex/analytics/rscript", "/data/data/analytics/rscript/")[[1]]
    setwd(g_working_directory);
    if(Sys.info()['sysname'] == "Linux") .libPaths(c("/data/data/analytics/packages/", .libPaths()))
    
    library(logging); library(mxnet)
    
    #Adding logger
    addHandler(handler = writeToFile, file = "log.txt"); basicConfig(level = "INFO")
    
    # Get data for training    
    train.x <- data.matrix(airquality[,c('Ozone', 'Solar.R', 'Wind')])
    train.x[is.na(train.x)] <- 0
    train.y <- as.numeric(ifelse(airquality[,c('Temp')] < 70, 0, 1))
    
    # Prepare the layers
    data <- mx.symbol.Variable("data")
    fc1 <- mx.symbol.FullyConnected(data, name="fc1", num_hidden=3 )
    act1 <- mx.symbol.Activation(fc1, name="relu1", act_type="relu") # "relu" tanh
    fc2 <- mx.symbol.FullyConnected(act1, name="fc2", num_hidden=3 )
    act2 <- mx.symbol.Activation(fc2, name="relu2", act_type="relu")
    fc3 <- mx.symbol.FullyConnected(act2, name="fc3", num_hidden=2)
    softmax <- mx.symbol.SoftmaxOutput(fc3, name="sm")
    
    # Get all available CPUs. http://mxnet.io/how_to/env_var.html
    core <- as.integer(min(7, parallel::detectCores()))
    cpu_devices = lapply(1:core, function(i) {mx.cpu(i)})
    loginfo(paste0('Available cores are: ', parallel::detectCores(), ', and using ', core))
    
    l_list_tuned_param <- list(act_type="relu", num.round=10, array.batch.size=8, learning.rate=0.07, momentum=0.9,  initializer=0.01, optimizer ="sgd")
    
    # Building model
    
    tryCatch(fit_dl <- mx.model.FeedForward.create(softmax, X=data.matrix(train.x), y = as.numeric(train.y), ctx=cpu_devices, num.round=l_list_tuned_param[['num.round']], array.batch.size=l_list_tuned_param[['array.batch.size']], learning.rate=l_list_tuned_param[['learning.rate']], momentum=l_list_tuned_param[['momentum']],  initializer=mx.init.uniform(l_list_tuned_param[['initializer']]), optimizer = l_list_tuned_param[['optimizer']],eval.metric=mx.metric.accuracy, array.layout = "rowmajor"), error = function(cond){logerror(cond); quit(save = "no", status = 0, runLast = F)})
    
    loginfo("Building complete")

**Comment by shivonkar** [comment]
The error message is attached

![prod2_error](https://user-images.githubusercontent.com/11136293/32714241-f28d2072-c872-11e7-825b-2fccdb8d5065.PNG)

**Comment by shivonkar** [comment]
Hi,
The github is updated

Thank you
Shiv

**Comment by shivonkar** [comment]
Here is solution.

First set the environment variable
Sys.setenv(MXNET_CPU_WORKER_NTHREADS = core); 

Use ctx= mx.cpu() depending upon cpu/gpu

It is working on both Windows and Linux

**Comment by anirudhacharya** [comment]
@shivonkar  close this issue if it is resolved.

@cjolivier01

**Comment by cjolivier01** [comment]
appears to be resolved — please reopen if necessary
